iot-hub/Samples/service/IotHubSasCredentialAuthenticationSample/Program.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using Azure.Core;
+using CommandLine;
+using Microsoft.Azure.Devices;
+using System;
+using System.Globalization;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace IotHubSasCredentialAuthenticationSample
+{
+    public class Program
+    {
+        /// <summary>
+        /// A sample to illustrate how to use Azure active directory for authentication to the IoT hub.
+        /// <param name="args">Run with `--help` to see a list of required and optional parameters.</param>
+        public static async Task Main(string[] args)
+        {            
+            // Parse application parameters
+            Parameters parameters = null;
+            ParserResult<Parameters> result = Parser.Default.ParseArguments<Parameters>(args)
+                .WithParsed(parsedParams =>
+                {
+                    parameters = parsedParams;
+                })
+                .WithNotParsed(errors =>
+                {
+                    Environment.Exit(1);
+                });
+
+            // Initialize SAS token credentials.
+            Console.WriteLine("Creating sas credential.");
+
+            TimeSpan timeToLive = TimeSpan.FromHours(1);
+            DateTime expiresOn = DateTime.UtcNow.Add(timeToLive);
+            IotHubSasCredential sasCredential = new IotHubSasCredential(
+                GenerateSasToken(parameters.HostName, parameters.SharedAccessKey, expiresOn),
+                expiresOn);
+
+            // This is how the credential can be updated in the IotHubSasCredential object whenever necessary. It is not 
+            // necessary to do before the token expires. The following just demonstrates how to do it.
+            // NOTE: Ensure to pass the new expiry.
+            DateTime newExpiresOn = DateTime.UtcNow.Add(timeToLive);
```

Comment:
As we have our own derived class, that class now exposes two overloads for Update. And if the user uses the one without the expiryDate parameter, we will be passing he wrong date downstream. This means we have to add some code to disable that method. The only way to do it is by throwing in a override as in this example draft PR - https://github.com/Azure/azure-iot-sdk-csharp/pull/1793/files

This made me re-think our approach here and want to know other opinions. I am not completely against the approach in the above PR as it keeps us from doing any parsing of the SAS token. But it does mean that there can be a small trial and error learning curve in case users don't read our documentation.

Reply:
Instead of taking in the DateTime in the constructor, is taking a ttl an option? That way we could use the ttl supplied in the constructor in case they don't provide one for the Update method.

Reply:
We could use the ttl to generate the DateTime, if that is what we and possibly the base credential class require.

Reply:
What if they change the ttl? Though chances are rare, I don't think we should make an assumption on that.

Reply:
The AMQP layer expects us to send the DateTime of expiry. So the initial design was to get it from user instead of parsing the token.

Reply:
And generating the DateTime can have some clock skew (it might be acceptable but it will happen). It will not be the same as the one on the SAS token.

Reply:
How will ttl provided at constructor time be valid/used for update operation?

Reply:
This is now fixed by using encapsulation instead of inheritance. Only one update method will be exposed now. See PR - https://github.com/Azure/azure-iot-sdk-csharp/pull/1793